internal/sidecar/service/volumereplication.go

```diff
@@ -0,0 +1,188 @@
+		&csiReplication.DisableVolumeReplicationRequest{
+			VolumeId:      req.VolumeId,
+			ReplicationId: req.ReplicationId,
+			Parameters:    req.Parameters,
+			Secrets:       data,
+		})
+	if err != nil {
+		klog.Errorf("Failed to enable volume replication: %v", err)
+		return nil, err
+	}
+
+	return &proto.DisableVolumeReplicationResponse{}, nil
+}
+
+// PromoteVolume fetches required information from kubernetes cluster and calls
+// CSI-Addons PromoteVolume service.
+func (rs *ReplicationServer) PromoteVolume(
+	ctx context.Context,
+	req *proto.PromoteVolumeRequest) (*proto.PromoteVolumeResponse, error) {
+	// Get the secrets from the k8s cluster
+	data, err := kube.GetSecret(ctx, rs.kubeClient, req.GetSecretName(), req.GetSecretNamespace())
+	if err != nil {
+		klog.Errorf("Failed to get secret %s in namespace %s: %v", req.GetSecretName(), req.GetSecretNamespace(), err)
+		return nil, status.Error(codes.InvalidArgument, err.Error())
+	}
+
+	_, err = rs.controllerClient.PromoteVolume(ctx,
+		&csiReplication.PromoteVolumeRequest{
+			VolumeId:      req.VolumeId,
+			ReplicationId: req.ReplicationId,
+			Parameters:    req.Parameters,
+			Secrets:       data,
+		})
+	if err != nil {
+		klog.Errorf("Failed to enable volume replication: %v", err)
+		return nil, err
+	}
+
+	return &proto.PromoteVolumeResponse{}, nil
+}
+
+// DemoteVolume fetches required information from kubernetes cluster and calls
+// CSI-Addons DemoteVolume service.
+func (rs *ReplicationServer) DemoteVolume(
+	ctx context.Context,
+	req *proto.DemoteVolumeRequest) (*proto.DemoteVolumeResponse, error) {
+	// Get the secrets from the k8s cluster
+	data, err := kube.GetSecret(ctx, rs.kubeClient, req.GetSecretName(), req.GetSecretNamespace())
+	if err != nil {
+		klog.Errorf("Failed to get secret %s in namespace %s: %v", req.GetSecretName(), req.GetSecretNamespace(), err)
+		return nil, status.Error(codes.InvalidArgument, err.Error())
+	}
+
+	_, err = rs.controllerClient.DemoteVolume(ctx,
+		&csiReplication.DemoteVolumeRequest{
+			VolumeId:      req.VolumeId,
+			ReplicationId: req.ReplicationId,
+			Parameters:    req.Parameters,
+			Secrets:       data,
+		})
+	if err != nil {
+		klog.Errorf("Failed to enable volume replication: %v", err)
+		return nil, err
+	}
+
+	return &proto.DemoteVolumeResponse{}, nil
+}
+
+// ResyncVolume fetches required information from kubernetes cluster and calls
+// CSI-Addons ResyncVolume service.
+func (rs *ReplicationServer) ResyncVolume(
+	ctx context.Context,
+	req *proto.ResyncVolumeRequest) (*proto.ResyncVolumeResponse, error) {
+	// Get the secrets from the k8s cluster
+	data, err := kube.GetSecret(ctx, rs.kubeClient, req.GetSecretName(), req.GetSecretNamespace())
+	if err != nil {
+		klog.Errorf("Failed to get secret %s in namespace %s: %v", req.GetSecretName(), req.GetSecretNamespace(), err)
+		return nil, status.Error(codes.InvalidArgument, err.Error())
+	}
+
+	_, err = rs.controllerClient.ResyncVolume(ctx,
```

Comment:
Resync should return success or failure based on that we need to take few actions to reconcile the CR or not.